traffic_ops/app/db/reencrypt/reencrypt.go

```diff
@@ -0,0 +1,370 @@
+	}
+
+	previousKey, err := readKey(*previousKeyLocation)
+	if err != nil {
+		die("reading previous-key: " + err.Error())
+	}
+
+	dbConfBytes, err := ioutil.ReadFile(*cfg)
+	if err != nil {
+		die("reading db conf '" + *cfg + "': " + err.Error())
+	}
+
+	pgCfg := Config{}
+	err = json.Unmarshal(dbConfBytes, &pgCfg)
+	if err != nil {
+		die("unmarshalling '" + *cfg + "': " + err.Error())
+	}
+
+	sslStr := "require"
+	if !pgCfg.SSL {
+		sslStr = "disable"
+	}
+	db, err := sqlx.Open("postgres", fmt.Sprintf("postgres://%s:%s@%s:%d/%s?sslmode=%s&fallback_application_name=trafficvault", pgCfg.User, pgCfg.Password, pgCfg.Hostname, pgCfg.Port, pgCfg.DBName, sslStr))
+	if err != nil {
+		die("opening database: " + err.Error())
+	}
+
+	tx, err := db.Begin()
+	if err != nil {
+		die(fmt.Sprintf("transaction begin failed %v %v ", err, tx))
+	}
+	defer tx.Commit()
+
+	if err = reEncryptSslKeys(tx, previousKey, newKey); err != nil {
+		tx.Rollback()
+		die("re-encrypting SSL Keys: " + err.Error())
+	}
+	if err = reEncryptUrlSigKeys(tx, previousKey, newKey); err != nil {
+		tx.Rollback()
+		die("re-encrypting URL Sig Keys: " + err.Error())
+	}
+	if err = reEncryptUriSigningKeys(tx, previousKey, newKey); err != nil {
+		tx.Rollback()
+		die("re-encrypting URI Signing Keys: " + err.Error())
+	}
+	if err = reEncryptDNSSECKeys(tx, previousKey, newKey); err != nil {
+		tx.Rollback()
+		die("re-encrypting DNSSEC Keys: " + err.Error())
+	}
+
+	if err = updateKeyFile(previousKey, *previousKeyLocation, newKey); err != nil {
+		die("updating the key file: " + err.Error())
+	}
+
+	fmt.Println("Successfully re-encrypted keys for SSL Keys, URL Sig Keys, URI Signing Keys, and DNSSEC Keys.")
+}
+
+type Config struct {
+	DBName   string `json:"dbname"`
+	Hostname string `json:"hostname"`
+	User     string `json:"user"`
+	Password string `json:"password"`
+	Port     int    `json:"port"`
+	SSL      bool   `json:"ssl"`
+}
+
+func readKey(keyLocation string) ([]byte, error) {
+	var keyBase64 string
+	keyBase64Bytes, err := ioutil.ReadFile(keyLocation)
+	if err != nil {
+		return []byte{}, errors.New("reading file '" + keyLocation + "':" + err.Error())
+	}
+	keyBase64 = string(keyBase64Bytes)
+
+	key, err := base64.StdEncoding.DecodeString(keyBase64)
+	if err != nil {
+		return []byte{}, fmt.Errorf("AES key cannot be decoded from base64: %w", err)
+	}
+
+	// verify the key works
+	_, err = aes.NewCipher(key)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return key, nil
+}
+
+func reEncryptSslKeys(tx *sql.Tx, previousKey []byte, newKey []byte) error {
+	rows, err := tx.Query("SELECT id, data FROM sslkey")
+	if err != nil {
+		return fmt.Errorf("querying: %w", err)
+	}
+	defer rows.Close()
+
+	sslKeyMap := map[int][]byte{}
+
+	for rows.Next() {
+		id := 0
+		var encryptedSslKeys []byte
+		if err = rows.Scan(&id, &encryptedSslKeys); err != nil {
+			return fmt.Errorf("getting SSL Keys: %w", err)
+		}
+		jsonKeys, err := util.AESDecrypt(encryptedSslKeys, previousKey)
```

Comment:
So as I was using this tool, I had a thought: what if we accidentally mixed up the keys -- e.g. passed the new key as `previous-key` and the old key as `new-key`? I think this might happily decrypt the data using the wrong key, which would result in garbage data, then re-encrypt the garbage data with the new key. I think we could reverse ourselves out of that situation if it happened, but we should probably include sanity-checks here between decrypting and re-encrypting. For example, making sure the decrypted data is actually JSON. Maybe we should try unmarshalling it into its struct and checking that the resulting struct isn't all empty?

Reply:
if you pass the wrong key in as the `new-key` then it gets this error:
`re-encrypting SSL Keys: reading SSL Keys: cipher: message authentication failed` when trying to decrypt

Reply:
probably worth putting those checks in anyway though. its too important of data to risk

Reply:
Oh, well that's good! I think when I tried this out I must've re-encrypted w/ the same key because I saved it off and reused it but it was overwritten. I'm like 0/3 today 🤕